### ISSUE TITLE
ML Feature Inventory: Complete Data Point Catalog, Umbrellas & Gap Analysis

### ISSUE THREAD

**Issue by holden** [razrfly/pokevestment#29]
## Purpose

Comprehensive inventory of every data point available for ML modeling, organized into human-readable umbrella categories for eventual UI/UX. Includes gap analysis of open issues to identify easy wins before we start running models.

This is a **reference document** — no code changes. It should live in the README or docs once finalized.

---

### What's Done (from #25, now closed)
All Priority 1-4 items from #25 are implemented and backfilled:
- Tournament data pipeline (4 new tables, ~600K+ deck card rows)
- Art type extraction (art_type, is_full_art, is_alternate_art)
- Set supply proxies (secret_rare_count, secret_rare_ratio, era)
- Language count (language_count on cards — currently 1-6, see #28)

### Easy Wins Still Open (should do before ML)

| Issue | What | Effort | Signal Value | Status |
|-------|------|--------|-------------|--------|
| **#28** | Fix language_count to include Japanese cards | 1-2 hrs | Medium | Open — Japanese ID mapping needs investigation |
| **#19 items 1-5** | Trivial derived features (days_since_release, set_card_count, set_legal, energy_cost_total, illustrator encoding) | Most are joins or computed at query time — 0 code for ML feature matrix assembly | High | Open but most are "free" at query time |
| **#19 items 6-8** | Species features (is_legendary, is_mythical, capture_rate, evolution_stage) | Already in DB via species table + card_dex_ids join — 0 code needed | High | Available now, just need to join |

**Recommendation**: #28 is the only actual code change needed. Everything in #19 is either already in the DB (join at query time) or computed (days_since_release = `Date.diff(today, set.release_date)`). These don't need new columns — the ML feature matrix assembly step handles them.

### Medium-Term Gaps (not blocking Model 1)

| Issue | What | Effort | Signal Value | Blocks |
|-------|------|--------|-------------|--------|
| **#15** | Historical price backfill | $10-20 + 1 day | Very High | Models 2-4 (trajectory, arbitrage) |
| **#21** | Regional/language pricing (Japanese market) | $10/mo + 1 day | Medium | Cross-market analysis |
| **#20 §3** | Sales volume data | Hard (paid API) | Very High | Confidence in price signals |
| **#20 §7** | Grading population (PSA/BGS census) | Medium (scraping) | Medium | Graded card valuation |

### Long-Term / Not Worth Doing Yet

| Issue | What | Why Wait |
|-------|------|----------|
| **#20 §8** | Social media sentiment | Extremely noisy signal, complex pipeline |
| **#9 §2** | Sealed product pricing | Separate product category, different model |
| **#20 §5** | Print run estimates | Not publicly available |

### Verdict

**We are ready for Model 1.** The only true easy win left is #28 (Japanese language_count fix). Everything else in #19 is available via joins at feature matrix assembly time. Issues #15, #21, #9 improve model quality but don't block the first model.

---

### Legend
- **Source**: Where the data lives in our DB
- **Type**: How ML should treat it (numeric, categorical, boolean, derived)
- **Coverage**: % of 22,754 cards that have this field populated
- **Signal**: Estimated predictive importance (from cross-sectional analysis in #11)

---

### Umbrella 1: Card Identity & Classification
*What the card IS — its fundamental properties.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 1 | `category` | cards.category | Categorical (3 values) | 100% | High | Pokemon / Trainer / Energy |
| 2 | `rarity` | cards.rarity | Categorical (65+ values) | 100% | **Very High** | Mega Hyper Rare avg $304 vs Common $4.67 (65x) |
| 3 | `art_type` | cards.art_type | Categorical (14 values) | 100% | **Very High** | Normalized rarity → art classification |
| 4 | `is_full_art` | cards.is_full_art | Boolean | 100% | High | Binary flag for any full/alt art |
| 5 | `is_alternate_art` | cards.is_alternate_art | Boolean | 100% | High | Covers illustration, special illustration, rainbow |
| 6 | `is_secret_rare` | cards.is_secret_rare | Boolean | 100% | High | $77 avg vs $12.48 non-secret (6.2x) |
| 7 | `suffix` | cards.suffix | Categorical (7 values) | 10% | High | ex, GX, V, VMAX, VSTAR, TAG TEAM-GX, etc. |
| 8 | `stage` | cards.stage | Categorical | 82% | Medium | Basic, Stage 1, Stage 2, MEGA, etc. |
| 9 | `hp` | cards.hp | Numeric | 85% | Low-Medium | Hit points (null for Trainers/Energy) |
| 10 | `trainer_type` | cards.trainer_type | Categorical | 12% | Low | Supporter, Item, Stadium, Tool (Trainers only) |
| 11 | `energy_type` | cards.energy_type | Categorical | 2% | Low | Basic energy type (Energy cards only) |
| 12 | `regulation_mark` | cards.regulation_mark | Categorical (A-H) | 34% | Medium | Modern cards only — era indicator |
| 13 | `first_edition` | cards.first_edition | Boolean | 100% | High | ~940 cards are 1st edition |
| 14 | `is_shadowless` | cards.is_shadowless | Boolean | 100% | High | Shadowless Base Set cards worth significantly more |
| 15 | `has_first_edition_stamp` | cards.has_first_edition_stamp | Boolean | 100% | Medium | Has 1st-edition in stamps |

### Umbrella 2: Combat & Gameplay Power
*How strong the card is in actual gameplay.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 16 | `attack_count` | cards.attack_count | Numeric (0-4) | 100% | Low-Medium | Number of attacks |
| 17 | `total_attack_damage` | cards.total_attack_damage | Numeric | 100% | Low | Sum of all attack damage |
| 18 | `max_attack_damage` | cards.max_attack_damage | Numeric | 100% | Low-Medium | Highest single attack |
| 19 | `has_ability` | cards.has_ability | Boolean | 100% | Low-Medium | Cards with abilities are often more playable |
| 20 | `ability_count` | cards.ability_count | Numeric | 100% | Low | Number of abilities |
| 21 | `weakness_count` | cards.weakness_count | Numeric | 100% | Low | Number of type weaknesses |
| 22 | `resistance_count` | cards.resistance_count | Numeric | 100% | Low | Number of type resistances |
| 23 | `retreat_cost` | cards.retreat_cost | Numeric | 70% | Low | Energy to retreat |
| 24 | `type_count` | card_types (count) | Numeric | 85% | Low | Number of energy types on card |
| 25 | `primary_type` | card_types (first) | Categorical | 85% | Medium | Fire, Water, Psychic, etc. |
| 26 | `energy_cost_total` | Derived from attacks JSONB | Numeric | 84% | Low | Sum of all attack energy costs (#19 item 4) |

### Umbrella 3: Pokemon Identity & Lore
*Which Pokemon is on the card and its in-game significance.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 27 | `generation` | cards.generation (via dex_ids) | Numeric (1-9) | 74% | **Very High** | Gen 1 avg $28.21 vs Gen 9 $1.14 (25x) |
| 28 | `is_legendary` | pokemon_species.is_legendary | Boolean | 74%* | High | $52.12 avg vs $14.09 (3.7x) |
| 29 | `is_mythical` | pokemon_species.is_mythical | Boolean | 74%* | Medium | $27.06 avg vs $14.09 (1.9x) |
| 30 | `is_baby` | pokemon_species.is_baby | Boolean | 74%* | Low | Baby Pokemon flag |
| 31 | `capture_rate` | pokemon_species.capture_rate | Numeric (3-255) | 74%* | Medium | Lower = rarer in games = potentially more popular |
| 32 | `species_color` | pokemon_species.color | Categorical | 74%* | Low | Body color classification |
| 33 | `species_habitat` | pokemon_species.habitat | Categorical | 26%** | Low | Only Gen 1-3 have habitats |
| 34 | `species_shape` | pokemon_species.shape | Categorical | 74%* | Low | Body shape |
| 35 | `growth_rate` | pokemon_species.growth_rate | Categorical | 74%* | Low | Experience growth rate |
| 36 | `evolution_stage_in_chain` | Derived from evolves_from_id chains | Categorical (base/mid/final) | 74%* | Medium | Final evolutions more popular (#19 item 8) |
| 37 | `evolves_from` | cards.evolves_from | String/Boolean | 35% | Low | Has evolution predecessor |

*\* Coverage is 74% because only Pokemon cards have dex_ids; Trainers/Energy are null — which is correct.*
*\*\* PokeAPI only has habitat for Gen 1-3.*

### Umbrella 4: Artist & Visual Appeal
*Who made the art and what kind of art treatment it received.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 38 | `illustrator` | cards.illustrator | Categorical (407 values) | 91% | **Very High** | Top artist avg $209 vs bottom <$1 (200x+) |
| 39 | `illustrator_frequency` | Derived (count per illustrator) | Numeric | 91% | Medium | Prolific vs rare illustrators |
| 40 | `has_image` | cards.image_url IS NOT NULL | Boolean | 96% | Low | Proxy for card data completeness |

*Note: art_type, is_full_art, is_alternate_art from Umbrella 1 are also visual — but they derive from rarity, so they're classified under Identity.*

### Umbrella 5: Set & Series Context
*Where and when the card was released — supply-side signals.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 41 | `set_id` | cards.set_id | Categorical (200 values) | 100% | Medium | Which set the card belongs to |
| 42 | `series_id` | sets.series_id | Categorical (21 values) | 100% | Medium | Which series the set belongs to |
| 43 | `era` | sets.era | Categorical (12 values) | 100% | High | wotc/ex/dp/bw/xy/sm/swsh/sv/mega/etc. |
| 44 | `days_since_release` | Derived: Date.diff(today, sets.release_date) | Numeric | 99% | **Very High** | Likely top-3 predictor. Older = more valuable. |
| 45 | `set_card_count_official` | sets.card_count_official | Numeric | 100% | Medium | Smaller sets = rarer pulls |
| 46 | `set_card_count_total` | sets.card_count_total | Numeric | 100% | Low | Total including secrets |
| 47 | `secret_rare_count` | sets.secret_rare_count | Numeric | 100% | Medium | Number of bonus/secret cards |
| 48 | `secret_rare_ratio` | sets.secret_rare_ratio | Numeric | 100% | Medium | Proportion of chase cards |
| 49 | `set_legal_standard` | sets.legal_standard | Boolean | 100% | Medium | Tournament legality affects demand |
| 50 | `set_legal_expanded` | sets.legal_expanded | Boolean | 100% | Low | Expanded format legality |
| 51 | `language_count` | cards.language_count | Numeric (1-7) | 100% | Medium | More languages = larger print run = more supply |

### Umbrella 6: Competitive / Tournament Demand
*How the card performs in competitive play — strongest short-term predictor.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 52 | `meta_share` | Derived: decks with card / total decks in period | Numeric (0-1) | Varies by window | **Very High** | Primary demand pressure signal |
| 53 | `tournament_appearances` | Derived: count distinct tournaments | Numeric | Varies | High | Breadth of competitive demand |
| 54 | `avg_copies_per_deck` | Derived: avg count when included | Numeric (1-4) | Varies | High | 4-of staple vs 1-of tech |
| 55 | `archetype_count` | Derived: distinct deck_archetype_ids | Numeric | Varies | High | Versatility across deck types |
| 56 | `avg_win_rate` | Derived: weighted avg wins/(wins+losses+ties) | Numeric (0-1) | Varies | Medium-High | Competitive strength signal |
| 57 | `meta_trend` | Derived: meta_share(7d) - meta_share(30d) | Numeric | Varies | **Very High** | Rising/falling demand momentum |
| 58 | `top_8_appearances` | Derived: count of standings where placing ≤ 8 | Numeric | Varies | High | Elite competitive performance |

*Note: Tournament features are computed over time windows (7d, 30d, 90d). Each window is effectively a separate feature, so 52-58 multiply by 2-3 windows = ~15-20 actual features.*

### Umbrella 7: Price & Market Signals
*Current and historical pricing data — both target variables and features.*

| # | Feature | Source | Type | Coverage | Signal | Notes |
|---|---------|--------|------|----------|--------|-------|
| 59 | `price_market` (TCGPlayer) | price_snapshots | Numeric | 42% | Target variable | TCGPlayer market price (USD) |
| 60 | `price_avg` (CardMarket) | price_snapshots | Numeric | 81% | Target variable | CardMarket average price (EUR) |
| 61 | `price_trend` (CardMarket) | price_snapshots | Numeric | 81% | Feature | CardMarket trend indicator |
| 62 | `price_avg1` | price_snapshots | Numeric | 81% | Feature | 1-day moving average |
| 63 | `price_avg7` | price_snapshots | Numeric | 81% | Feature | 7-day moving average |
| 64 | `price_avg30` | price_snapshots | Numeric | 81% | Feature | 30-day moving average |
| 65 | `price_momentum_7d` | Derived: price_avg1 / price_avg7 | Numeric | 81% | High | Short-term momentum |
| 66 | `price_momentum_30d` | Derived: price_avg1 / price_avg30 | Numeric | 81% | High | Medium-term momentum |
| 67 | `tcg_cm_spread` | Derived: price_market / price_avg | Numeric | 42% | Medium | Cross-market price divergence |
| 68 | `price_low` | price_snapshots | Numeric | 84% | Low | Floor price |
| 69 | `price_high` | price_snapshots | Numeric | 84% | Low | Ceiling price |
| 70 | `price_spread` | Derived: (high - low) / mid | Numeric | 84% | Medium | Market uncertainty/volatility |

*Note: Features 62-66 are available TODAY from CardMarket's rolling averages — we don't need to wait for our own time-series to accumulate for basic momentum signals.*

---

### By Umbrella

| # | Umbrella | Features | Key Signal | UI Display Priority |
|---|----------|----------|-----------|-------------------|
| 1 | **Card Identity & Classification** | 15 | Rarity, art type, secret rare | High — users understand "it's a secret rare" |
| 2 | **Combat & Gameplay Power** | 11 | Weak individually, moderate in aggregate | Low — not very interesting to investors |
| 3 | **Pokemon Identity & Lore** | 11 | Generation, legendary status | High — "it's a Gen 1 legendary" resonates |
| 4 | **Artist & Visual Appeal** | 3 | Illustrator premium | Medium — interesting alpha, niche audience |
| 5 | **Set & Series Context** | 11 | Age, era, language count | Medium — "vintage set" is intuitive |
| 6 | **Competitive / Tournament Demand** | 7 (×3 windows) | Meta share, meta trend | **Very High** — strongest short-term predictor |
| 7 | **Price & Market Signals** | 12 | Momentum, spread | High — direct price action |

### Totals
- **70 base features** across 7 umbrellas
- **~85-90 actual features** when tournament time windows are expanded
- **42 features** available at 100% coverage (no nulls)
- **19 features** require species join (null for non-Pokemon cards — correct behavior)
- **7 tournament features** are derived at query time from tournament tables

### By Signal Strength (estimated)

| Tier | Features | Notes |
|------|----------|-------|
| **Very High** | rarity/art_type, generation, days_since_release, illustrator, meta_share, meta_trend, price_momentum | These 7 likely explain 80%+ of variance |
| **High** | is_legendary, is_secret_rare, suffix, era, first_edition, is_shadowless, tournament_appearances, avg_copies_per_deck, archetype_count | Strong secondary signals |
| **Medium** | secret_rare_ratio, capture_rate, language_count, regulation_mark, set_card_count, hp, stage, primary_type, price_spread, evolution_stage | Moderate predictive power |
| **Low** | Most combat stats, species_color/shape/habitat, energy_type, trainer_type, growth_rate | Individually weak but may contribute in ensemble |

---

### For the ML Pipeline
The model should weight features **individually** (not by umbrella). XGBoost/LightGBM does this automatically via feature importance. The umbrellas are for human interpretation only.

### For the UI/UX
When explaining "why is this card undervalued?", group the top contributing features under their umbrella:

```
🟢 STRONG BUY — Pikachu VMAX (Secret)

Value Drivers:
  📦 Card Identity: Secret rare (6.2x premium), VMAX suffix (3.4x)
  🧬 Pokemon Lore: Generation 1 (25x vs Gen 9), iconic species
  🎨 Artist: HYOGONOSUKE (top-tier illustrator, 2.1x premium)
  📅 Set Context: 1,800 days old (vintage premium building)
  🏆 Tournament: 12% meta share, appearing in 4 archetypes
```

The umbrellas give users a mental model: "this card is valuable because of WHAT it is + WHO made it + WHEN it came from + HOW it's used in competition."

### For Future Feature Engineering
When we add new data sources, classify them into existing umbrellas or create new ones:

| Future Feature | Umbrella | Source |
|---------------|----------|--------|
| Graded population (PSA census) | New: **Scarcity & Grading** | PSA API / scraping |
| Historical price backfill | Price & Market Signals | PokemonPriceTracker |
| Japanese market pricing | Price & Market Signals | PokemonPriceTracker |
| Social media mentions | New: **Hype & Sentiment** | Reddit/YouTube APIs |
| Sealed product prices | New: **Sealed Market** | Various |
| Sales volume | Price & Market Signals | TCGPlayer affiliate |

---

## Part 5: Open Issue Disposition

| Issue | Action | Rationale |
|-------|--------|-----------|
| **#9** | Keep open | Graded + sealed + historical research — Phase 2+ |
| **#11** | Close or update | Daily sync is running. Feature extraction done. Most items complete. |
| **#12** | Keep open | ML strategy — we're about to execute on this |
| **#14** | Keep open | Roadmap/meta issue — tracks the overall vision |
| **#15** | Keep open | Historical backfill — accelerates trajectory models |
| **#19** | Close | All items are either implemented (#25) or available via joins. No code needed. |
| **#20** | Keep open | Data gaps tracking — long-term research |
| **#21** | Keep open | Regional pricing — Phase 2+ |
| **#26** | Keep open | This is the next major implementation issue (models + UI) |
| **#28** | Keep open | Japanese language_count fix — quick win |

**Suggested closures**: #19 (superseded by #25 implementation + this inventory), #11 (substantially complete)

---

## Related Issues
- #25 (Schema normalization — now closed, implemented all P1-P4)
- #26 (Investment Intelligence Platform — next major milestone)
- #19 (Feature Engineering — can close, this issue supersedes)
- #12 (ML Strategy — execution starts now)
- #14 (New Set Advisor roadmap)
- #15, #20, #21 (Data gaps — tracked for future improvement)
- #28 (Japanese language_count fix — only remaining easy win)

**Comment by holden** [comment]
Closing — this was a reference document cataloging all ML features. The feature inventory is complete: ~85-90 features across 7 umbrellas are implemented in the feature matrix (`FeatureMatrix`, `PriceFeatures`, `TournamentFeatures`). Gap analysis is done. All actionable items from this issue have been implemented or tracked elsewhere. See #56 for current roadmap.
